pyamg/classical/cr.py

```diff
@@ -3,290 +3,189 @@
-            return splitting.ravel()
-
-        # now add points
-        #
-        # update limit on additions to splitting (C)
-        if alpha < 1e-13:
-            alpha = 0.25
-        else:
-            alpha = (1-alphai) * alpha + alphai * (1/gamma)
-
-        nCmax = np.ceil(alpha * n)
-
-        L = np.ceil(G * E / E.max()).ravel()
-
-        binid = G
-
-        # add whole bins (and t-depth nodes) at a time
-        # u = np.zeros((n, 1))
-        # TODO This loop may never halt...
-        #      Perhaps loop over nC < nCmax and binid > 0 ?
-        while nC < nCmax:
-            if delta > 0:
-                raise NotImplementedError
-            if tdepth != 1:
-                raise NotImplementedError
-
-            (roots,) = np.where(L == binid)
-
-            for root in roots:
-                if L[root] >= 0:
-                    cols = S[root, :].indices
-                    splitting[root] = 1    # add roots
-                    nC += 1
-                    L[cols] = -1
-            binid -= 1
-
-            # L[troots] = -1          # mark t-rings visited
-            # u[:]=0.0
-            # u[roots] = 1.0
-            # for depth in range(0,tdepth):
-            #     u = np.abs(S) * u
-            # (troots,tmp) = np.where(u>0)
-
-    return splitting.ravel()
-
-
-def _CRsweep(A, Findex, Cindex, nu, thetacr, method):
+def _CRsweep(A, B, Findex, Cindex, nu, thetacr, method):
 
     n = A.shape[0]    # problem size
     numax = nu
     z = np.zeros((n,))
-    e = np.ones((n,))
+    e = deepcopy(B[:,0])
     e[Cindex] = 0.0
     enorm = norm(e)
     rhok = 1
+    it = 0
 
-    for it in range(1, numax+1):
-
+    while True:
         if method == 'habituated':
             gauss_seidel(A, e, z, iterations=1)
             e[Cindex] = 0.0
         elif method == 'concurrent':
             gauss_seidel_indexed(A, e, z, indices=Findex, iterations=1)
         else:
-            raise NotImplementedError('method not recognized: need \
-                                       habituated or concurrent')
+            raise NotImplementedError('method not recognized: need habituated or concurrent')
 
         enorm_old = enorm
         enorm = norm(e)
         rhok_old = rhok
         rhok = enorm / enorm_old
+        it += 1
 
-        # criteria 1
-        if (abs(rhok - rhok_old) / rhok < 0.1) and (it >= nu):
-            return rhok, e
+        # criteria 1 -- at least nu iters, relative change in CF is small (<0.1)
+        if ( (abs(rhok - rhok_old) / rhok) < 0.1) and (it >= nu):
+            break
+        # criteria 2 -- fast convergence 
+        elif rhok < 0.1 * thetacr:
+            break
 
-        # criteria 2
-        if rhok < 0.1 * thetacr:
-            return rhok, e
+    return rhok, e
 
 
-def CRalpha(A, method='habituated', nu=3, thetacr=0.7, thetacs=[0.3, 0.5],
-            maxiter=20):
-    """
+def CR(A, method='habituated', B=None, nu=3, thetacr=0.5,
+        thetacs='auto', maxiter=20, verbose=False):
```

Comment:
Maybe include 2 small unit tests that compare the CF splitting to a stored CF splitting.